src/relay/libp2p_relay_server.erl

```diff
@@ -225,7 +225,17 @@ init_relay(#state{tid = TID, banlist = Banlist}) ->
     lager:debug("init relay for swarm ~p", [libp2p_swarm:name(Swarm)]),
     case libp2p_peerbook:random(Peerbook,[SwarmPubKeyBin | Banlist],
                                 fun(P) ->
-                                        libp2p_peer:has_public_ip(P) andalso
+                                        %% this is complex and restrictive, but there's a hard to
+                                        %% track down bug in the relay logic that sometimes allows
+                                        %% ipv4 addresses to be listed within the listen addrs.
+                                        %% this can cause relay establishment problems because the
+                                        %% relays only work if they're in the correct order, which
+                                        %% is impossible to guarantee.  so we work around by
+                                        %% narrowing the relay pool to stuff that we know is likely
+                                        %% to work, or to at least fail immediately.
+                                        ListenAddrs = libp2p_peer:listen_addrs(P),
```

Comment:
Do we no longer need to check whether the peer `has_public_ip`?

Reply:
we still check it